Add code_format_template to template locks

chammp opened the pull request:
<!--
  You are amazing! Thanks for contributing to our project!
  Please, DO NOT DELETE ANY TEXT from this template! (unless instructed).
-->

## Proposed change
Adds a configuration option `code_format_template` (Template resolving to regexp) to Template Lock entities. This allows the locks to require a code to lock/unlock as other types of locks also do. The code is validated against the given format as per the default Lock implementation, and the value entered by the users is passed to the configured lock/unlock scripts as template variable `code`, where additional validation (also dynamic) can be performed.

The actual code format is dynamic. The frontend queries the backend for the current entity config when lock operations are performed so that UI knows whether to show the code input dialog in `callProtectedLockService`:
https://github.com/home-assistant/frontend/blob/be26f8bc24809e7db906b4b58ea001834b639392/src/data/lock.ts#L58-L66
This ensures that the frontend does not assume static code formats for lock entities.

## Type of change
<!--
  What type of change does your PR introduce to Home Assistant?
  NOTE: Please, check only 1! box!
  If your PR requires multiple boxes to be checked, you'll most likely need to
  split it into multiple PRs. This makes things easier and faster to code review.
-->

- [ ] Dependency upgrade
- [ ] Bugfix (non-breaking change which fixes an issue)
- [ ] New integration (thank you!)
- [x] New feature (which adds functionality to an existing integration)
- [ ] Deprecation (breaking change to happen in the future)
- [ ] Breaking change (fix/feature causing existing functionality to break)
- [ ] Code quality improvements to existing code or addition of tests

## Additional information
<!--
  Details are important, and help maintainers processing your PR.
  Please be sure to fill out additional details, if applicable.
-->

- This PR fixes or closes issue: fixes #
- This PR is related to issue: 
- Link to documentation pull request: https://github.com/home-assistant/home-assistant.io/pull/32791

## Checklist
<!--
  Put an `x` in the boxes that apply. You can also fill these out after
  creating the PR. If you're unsure about any of them, don't hesitate to ask.
  We're here to help! This is simply a reminder of what we are going to look
  for before merging your code.
-->

- [x] The code change is tested and works locally.
- [x] Local tests pass. **Your PR cannot be merged unless tests pass**
- [x] There is no commented out code in this PR.
- [x] I have followed the [development checklist][dev-checklist]
- [x] I have followed the [perfect PR recommendations][perfect-pr]
- [x] The code has been formatted using Ruff (`ruff format homeassistant tests`)
- [x] Tests have been added to verify that the new code works.

If user exposed functionality or configuration variables are added/changed:

- [ ] Documentation added/updated for [www.home-assistant.io][docs-repository]

If the code communicates with devices, web services, or third-party tools:

- [ ] The [manifest file][manifest-docs] has all fields filled out correctly.  
      Updated and included derived files by running: `python3 -m script.hassfest`.
- [ ] New or updated dependencies have been added to `requirements_all.txt`.  
      Updated by running `python3 -m script.gen_requirements_all`.
- [ ] For the updated dependencies - a link to the changelog, or at minimum a diff between library versions is added to the PR description.
- [ ] Untested files have been added to `.coveragerc`.

<!--
  This project is very active and we have a high turnover of pull requests.

  Unfortunately, the number of incoming pull requests is higher than what our
  reviewers can review and merge so there is a long backlog of pull requests
  waiting for review. You can help here!
  
  By reviewing another pull request, you will help raise the code quality of
  that pull request and the final review will be faster. This way the general
  pace of pull request reviews will go up and your wait time will go down.
  
  When picking a pull request to review, try to choose one that hasn't yet
  been reviewed.

  Thanks for helping out!
-->

To help with the load of incoming pull requests:

- [ ] I have reviewed two other [open pull requests][prs] in this repository.

[prs]: https://github.com/home-assistant/core/pulls?q=is%3Aopen+is%3Apr+-author%3A%40me+-draft%3Atrue+-label%3Awaiting-for-upstream+sort%3Acreated-desc+review%3Anone+-status%3Afailure

<!--
  Thank you for contributing <3

  Below, some useful links you could explore:
-->
[dev-checklist]: https://developers.home-assistant.io/docs/development_checklist/
[manifest-docs]: https://developers.home-assistant.io/docs/creating_integration_manifest/
[quality-scale]: https://developers.home-assistant.io/docs/integration_quality_scale_index/
[docs-repository]: https://github.com/home-assistant/home-assistant.io
[perfect-pr]: https://developers.home-assistant.io/docs/review-process/#creating-the-perfect-pr

chammp committed:
Add code_format to template locks

chammp committed:
Replace code_format with code_format_template

chammp committed:
Add test case for template eval to None

emontnemery commented on homeassistant/components/template/lock.py:
```diff
@@ -47,6 +49,7 @@
         vol.Required(CONF_LOCK): cv.SCRIPT_SCHEMA,
         vol.Required(CONF_UNLOCK): cv.SCRIPT_SCHEMA,
         vol.Required(CONF_VALUE_TEMPLATE): cv.template,
+        vol.Optional(CONF_CODE_FORMAT_TEMPLATE): cv.template,
```
Why is this a template instead of a fixed string? Can you give some example of where it's useful to dynamically change the code format?

chammp replied:
This seemed to be the most straightforward way to support "require no code at all to lock, but a certain code to unlock". Or a well-known short code to lock, but a long (possibly dynamic) one to unlock.
From a system design perspective, if so desired it is always possible to go to a "secure" state (locked) and you need knowledge to go to an "insecure" (unlocked) state. Or have it the other way around if the meaning of "secure" and "insecure" is swapped in a user's context.

This works neatly with the frontend without introducing new attributes like [Template Alarm's code_arm_required](https://www.home-assistant.io/integrations/alarm_control_panel.template/#code_arm_required).

emontnemery replied:
OK I see. That makes the PR work around limitations in the lock entity IMHO, something which is not wanted. Please replace the `code_format_template` with a fixed `code_format` and hit the "Ready for review"-button when done.

chammp replied:
The long story is that I basically wanted to a) virtually secure certain switches/complex setups and b) have something similar to a switch that properly supports in-between-states. Meaning something complex is turned on (which I don't want to happen by accident and not necessarily by anyone walking through our our hallway), and while this is turning on, which can take quite some time, the UI controls display an intermediate state (like the lock that is "unlocking" and then is "unlocked"). The opposite is not true, so while I whish the turning on to be secured, turning off can happen anytime without securing it.

That drove me to template locks, which currently don't support codes at all, and make this available _and_ "templateable." 

As far as I see it, the main question is: does any existing code rely on the (undocumented, AFAIK) fact that a code format property of a lock never changes?  
I guess the answer is no, because there are places where this can obviously occur, Matter being one example. If no one relies on constant code format, why not make it optionally dynamic for template locks (that target advanced users only) as well? And the format can be based on anything, really, not only locks' state.

Another application I have in mind is a dynamic automatic parental control where one can use something while parents are at home, but need to enter a code if they are not. (All this applies to control surfaces that are accessible to anyone in the house).

emontnemery replied:
OK, I see, your use case is essentially about (ab)using lock entities for things which are not actually locks.

I think what you want to do is covered here https://github.com/home-assistant/frontend/discussions/8905, although there's no solution in that discussion. My personal opinion is that we should allow users to set a flag on any entity marking it as requiring confirmation, optionally with a PIN. Frontend and voice assistants should be aware of that flag and show a confirmation dialog / verbally ask for confirmation.

For this PR, I would suggest to remove the templating to get the base functionality - adding code format support to template locks merged. A follow-up PR could support templates where we can continue the discussion.

chammp replied:
It's not more abuse than [using a light to control home theater volume](https://www.home-assistant.io/integrations/light.template/#theater-volume-control) 😉. I always fancy the template-integration to be something special that exactly enables flexibility like this. It's kind of in its DNA.


I am also on board with adding UI-security measures (UI also meaning interactions by voice assistants) to all interactable entities, also with custom messages as described in the linked frontend issue. However, there are three considerations with this idea and your proposed next steps for this PR (all IMO of course):

1. This property (pin required/not required) needs to be templateable; user (or role) specific settings would be nice as I wouldn't bother myself to enter a PIN to enable A/C, for instance, because I usually sometimes kind of tend to know what I do 😉 
2. The place where the check happens should be customizable. This means that I expect the actual check (after the _requirement_ of a PIN/Code has been determined) to be offloaded to something else to be able to also support changing/rolling codes. Only supporting a static setting here would again severely limit flexibility.
3. Introducing a `code_format` for template locks now would solve about half of my issues, but _if_ later we come to the conclusion that it indeed should be templatable, it would either not fit to the template entities config convention (`_template` suffix) or it would create a second property and deprecate the old one and we'd have to think about procedure when both are set and whatnot...


You think a discussion on the discord would provide more insights into what others think about this whole ordeal? Would it even make a difference?

emontnemery commented on homeassistant/components/template/lock.py:
```diff
@@ -129,24 +135,58 @@ def _update_state(self, result):
+    @property
+    def code_format(self) -> str | None:
+        """Regex for code format or None if no code is required."""
+        if self._code_format is not None:
+            return self._code_format
+
+        return super().code_format
```
We don't need to call super here
```suggestion
        return self._code_format
```

chammp committed:
Apply suggestion to not call super()

Co-authored-by: Erik Montnemery <erik@montnemery.com>

chammp commented:
CI failure is unrelated to change but a rate limit imposed when uploading codecov report, see
![grafik](https://github.com/home-assistant/core/assets/57918757/09a353e6-7fd6-4c69-bb05-99502e6fd4b3)

emontnemery reviewed (changes requested):
The change is accepted, however:

- Add negative tests which show what happens if the template has a syntax error or raises an error during rendering? Will the result be that we accept any code or no code at all or something else? The behavior should also be clear in the documentation since it's a security concern
- Confirm in the PR description how frontend behaves when the code_format attribute changes
- Open a documentation PR and link to this PR

chammp committed:
Add more negative tests

chammp commented:
> * Add negative tests which show what happens if the template has a syntax error or raises an error during rendering? Will the result be that we accept any code or no code at all or something else? The behavior should also be clear in the documentation since it's a security concern

I added two tests that show the behavior

1. In case of template render errors (for instance divide by zero): here HA would accept _all_ codes (this is not a security issue if you follow the recommendation in the docs to always check the actual code in your actions)
2. In case of invalid rendered regexps; here basically no lock operation is possible because no entered code can match an invalid regexp.

Test for syntax errors in the template are already present (same behavior as when there are syntax errors in `value_template`).

> * Confirm in the PR description how frontend behaves when the code_format attribute changes

I added some lines to the PR; hopefully that's sufficient.

> * Open a documentation PR and link to this PR

Done and linked.

chammp commented:
CI test failure of most recent run is unrelated.

emontnemery commented:
> In case of template render errors (for instance divide by zero): here HA would accept all codes (this is not a security issue if you follow the recommendation in the docs to always check the actual code in your actions)

Right, but if the template is complicated, it could fail to render sometimes in ways the user didn't expect? Would it be better to fail all service calls then? Or it doesn't matter, because as you say, the point of the code_format_template is to give hints to frontend, not to actually validate the code?

emontnemery reviewed (approved):
I think this is good to go now, thanks @chammp 👍 

Before merging, I'd like a second opinion on if the error handling is sane.
Maybe it would be easier to understand for users if all codes are rejected both if the regex is invalid / can't match and if the template has errors?

emontnemery commented:
@chammp I discussed the PR with another developer. Please change it such that a failure to render the template also means no code will be accepted. That will make the behavior easier to document and easier for the user to understand.

edenhaus commented:
In my opinion any error is unexpected and we should raise to inform the user about it instead of silently ignoring and accepting any code.

chammp committed:
Handle template render errors

chammp commented:
Template render errors are now raised before any state changes occur and are displayed in UI (for my bogous template `"{{ 1/0 }}"`):

Dashboard:
![Dashboard Service Call Error](https://github.com/home-assistant/core/assets/57918757/038c761c-a478-4e16-badc-4d8ed181c8b9)

Dev Tools Service Call:
![Dev Tools Service Call Error](https://github.com/home-assistant/core/assets/57918757/8acb58c8-27e2-4b9b-82ea-5d356238e8a5)

And in Logs:
![Service Call Error in Logs](https://github.com/home-assistant/core/assets/57918757/df449b65-aea2-49b7-9223-e7129a203ba2)

emontnemery commented on homeassistant/components/template/lock.py:
```diff
@@ -130,24 +137,67 @@ def _update_state(self, result):
+
+    def _raise_template_error_if_available(self):
+        if self._code_format_template_error is not None:
+            raise ServiceValidationError(self._code_format_template_error)
```
Nice!
Please customize this so it's clear what template errored, use a translated message with placeholders for the name of the failing template and the error message in the _code_format_template_error

chammp replied:
Now:
![grafik](https://github.com/home-assistant/core/assets/57918757/ac9e97f5-314c-446c-8146-69891a6c94c8)

chammp committed:
Better error message

edenhaus commented on tests/components/template/test_lock.py:
```diff
@@ -14,13 +20,15 @@
         "data_template": {
             "action": "lock",
             "caller": "{{ this.entity_id }}",
+            "code": "{{ code }}",
```
Why is this change required?

chammp replied:
Because tests fail otherwise 🤷‍♂️ 😉 

I chimed into the way the other tests in this file use this config preset to assess whether actions are called with correct parameters.

edenhaus replied:
That's not an answer.

The code is needed only by your newly added tests. By reverting your changes in the test file, I can run all tests successfully.

chammp replied:
Seriously? If "I chimed into the way the other tests in this file use this config preset to assess whether actions are called with correct parameters." is not an answer then I don't know what is.

I guess you know how the tests work in this file, so what's the matter here? I honestly don't get it 🤷

edenhaus replied:
Providing a `code` is optional for the lock service calls.
I would leave the current service as they are (meaning I would revert this change)
If your newly added tests need a code, they add it only for these particular tests.
Could you also add a test in which a code_format is configured but no code is passed on the service call?

emontnemery replied:
I agree with @edenhaus, add a new config dict `OPTIMISTIC_LOCK_CONFIG_WITH_CODE` and use that in the new tests + add the negative test he requested.

Once that's sorted, this PR is good to go 👍

chammp replied:
That test that @edenhaus mentioned is already there; however I made that section a little more complete. Specific lock config for new tests is there as well.

edenhaus commented:
Drafting as we required test changes

chammp committed:
Add custom test lock config for code format

chammp committed:
Merge branch 'dev' into add-code-to-template-locks

chammp committed:
Add type hints from upstream

chammp commented:
Recent CI failures are unrelated:

![grafik](https://github.com/home-assistant/core/assets/57918757/fa9090f4-d3ec-4f8d-9858-b914d23c5a34)

frenck committed:
Merge branch 'dev' into add-code-to-template-locks

emontnemery committed:
Merge branch 'dev' into add-code-to-template-locks

emontnemery reviewed (approved):
LGTM, thanks @chammp 👍
